include/private/net/libssl.hpp

```diff
@@ -0,0 +1,331 @@
+
+        Links
+        -----
+
+        .. [1] https://wiki.openssl.org/index.php/Manual:SSL_CTX_free(3)
+        .. [2] https://github.com/openssl/openssl/issues/2165
+    */
+
+    /*!
+        \brief Creates a `SSL *` from the wrapped `SSL_CTX *`.
+
+        \param hostname Hostname for which to create the `SSL *`. This would be
+        the hostname for which you want to use SNI.
+
+        \param logger The logger to use to emit log messages.
+
+        \return An error on failure (the returned object will evaluate to false
+        in a boolean context) and a `SSL *` on success.
+    */
+    template <MK_MOCK(SSL_new)>
+    ErrorOr<SSL *> get_client_ssl(std::string hostname, Var<Logger> logger) {
+        assert(ctx_ != nullptr);
+        SSL *ssl = SSL_new(ctx_);
+        if (ssl == nullptr) {
+            logger->warn("ssl: SSL_new failed");
+            return SslNewError();
+        }
+        SSL_set_tlsext_host_name(ssl, hostname.c_str());
+        return ssl;
+    }
+
+    /// Destructor.
+    ~Context() {
+        assert(ctx_ != nullptr);
+        SSL_CTX_free(ctx_);
+    }
+
+    /*!
+        \brief Factory that creates a shared `SSL_CTX *` wrapper.
+
+        \param ca_bundle_path Path to the CA bundle to be used to verify
+        the servers we're connecting to. If it is empty and we're linking
+        with libressl, we will use the builtin CA bundle that is shipped
+        along with measurement-kit. Otherwise, we will fail.
+
+        \param logger Logger for printing log messages.
+
+        \return An error on failure (the returned object will evaluate to
+        false in a boolean context) or a `Var` shared pointer wrapping a
+        Context instance on success.
+    */
+    template <MK_MOCK(SSL_CTX_new), MK_MOCK(SSL_CTX_load_verify_locations)
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+                                          ,
+              MK_MOCK(SSL_CTX_load_verify_mem)
+#endif
+              >
+    static ErrorOr<Var<Context>> make(std::string ca_bundle_path,
+                                      Var<Logger> logger) {
+        // Implementation note: we need to initialize libssl early otherwise
+        // most of the functions of the library will not work properly.
+        init_libssl_once(logger);
+        logger->debug("ssl: creating SSL_CTX with bundle: '%s'",
+                      ca_bundle_path.c_str());
+        SSL_CTX *ctx = SSL_CTX_new(SSLv23_client_method());
+        if (ctx == nullptr) {
+            logger->warn("ssl: failed to create SSL_CTX");
+            return SslCtxNewError();
+        }
+        /*
+         * Implementation note:
+         *
+         * We use SSLv23_client_method() because it is the most general method
+         * that applications SHOULD use according to the manual[1] and we
+         * disable by default deprecated protocols. We MAY want to enable them
+         * in the OONI's web_connectivity test so to reach more websites.
```

Comment:
We actually _do_ that. Maybe I can update this comment as part of merging the PR.